packages/utils/src/eslint-utils/deepMerge.ts

```diff
@@ -22,28 +22,27 @@ export function deepMerge(
   // get the unique set of keys across both objects
   const keys = new Set(Object.keys(first).concat(Object.keys(second)));
 
-  return Array.from(keys).reduce<ObjectLike>((acc, key) => {
-    const firstHasKey = key in first;
-    const secondHasKey = key in second;
-    const firstValue = first[key];
-    const secondValue = second[key];
+  return Object.fromEntries(
+    Array.from(keys).map(key => {
+      const firstHasKey = key in first;
+      const secondHasKey = key in second;
+      const firstValue = first[key];
+      const secondValue = second[key];
 
-    if (firstHasKey && secondHasKey) {
-      if (isObjectNotArray(firstValue) && isObjectNotArray(secondValue)) {
-        // object type
-        acc[key] = deepMerge(firstValue, secondValue);
-      } else {
-        // value type
-        acc[key] = secondValue;
-      }
-    } else if (firstHasKey) {
-      acc[key] = firstValue;
-    } else {
-      acc[key] = secondValue;
-    }
-
-    return acc;
-  }, {});
+      return [
+        key,
+        firstHasKey && secondHasKey
+          ? isObjectNotArray(firstValue) && isObjectNotArray(secondValue)
+            ? // object type
+              deepMerge(firstValue, secondValue)
+            : // value type
+              secondValue
+          : firstHasKey
+            ? firstValue
+            : secondValue,
+      ];
+    }),
+  );
```

Comment:
Consider keeping the diff simpler like so:
```suggestion
      let value: unknown;

      if (firstHasKey && secondHasKey) {
        if (isObjectNotArray(firstValue) && isObjectNotArray(secondValue)) {
          // object type
          value = deepMerge(firstValue, secondValue);
        } else {
          // value type
          value = secondValue;
        }
      } else if (firstHasKey) {
        value = firstValue;
      } else {
        value = secondValue;
      }

      return [key, value];
```